Examples testing doesn't work with Sybil 6.1.0

### What happened?

When upgrading from Sybil 6.0.3 to 6.1.0 we get the following error:

~~~
Example at /home/luca/devel/repo-config/src/frequenz/repo/config/pytest/examples.py, line 17, column 1 did not evaluate as expected:
Pylint validation failed for code example:

# Generated auto-imports for code example
# pylint: disable=unused-import,wildcard-import,unused-wildcard-import
import ast
import os
import subprocess
import textwrap
from pathlib import Path
from typing import Any
from sybil import Example
from sybil.evaluators.python import pad
from sybil.parsers.myst import CodeBlockParser
from frequenz.repo.config.pytest.examples import *
# pylint: enable=unused-import,wildcard-import,unused-wildcard-import

from frequenz.repo.config.pytest import examples
from sybil import Sybil

pytest_collect_file = Sybil(**examples.get_sybil_arguments()).pytest()
```
Output: ************* Module config.pytest.examples
src/frequenz/repo/config/pytest/examples.py:22:1: E0001: Parsing failed: 'invalid syntax (config.pytest.examples, line 22)' (syntax-error)

examples.py:17: SybilFailure
~~~

### What did you expect instead?

It works.

### Affected version(s)

v0.9.x

### Affected part(s)

Build script, CI, dependencies, etc. (part:tooling), Tools to configure pytest (part:pytest)

### Extra information

When printing the generated source, this is what's being passed to pylint:

~~~py
# Generated auto-imports for code example
# pylint: disable=unused-import,wildcard-import,unused-wildcard-import
import ast
import os
import subprocess
import textwrap
from pathlib import Path
from typing import Any
from sybil import Example
from sybil.evaluators.python import pad
from sybil.parsers.myst import CodeBlockParser
from frequenz.repo.config.pytest.examples import *
# pylint: enable=unused-import,wildcard-import,unused-wildcard-import

from frequenz.repo.config.pytest import examples
from sybil import Sybil

pytest_collect_file = Sybil(**examples.get_sybil_arguments()).pytest()
```
~~~

The trailing \`\`\` is clearly causing the issue (it is line 22).

I'm not sure if we were misusing Sybil or if there is a regression in Sybil, it probably needs to be investigated.

If it is an issue in Sybil, we might want to deal with it ourselves for now, otherwise we'll need to pin sybil in all downstream projects too.

From [Sybil 6.1.0 changelog](https://github.com/simplistix/sybil/blob/master/CHANGELOG.rst#610-22-apr-2024):

> * Add support for lexing nested fenced codeblocks in markdown.
> * Add support for tilde-delimited codeblocks in markdown.

Probably related to this issue.

--- a/RELEASE_NOTES.md
+++ b/RELEASE_NOTES.md
@@ -44,7 +44,7 @@
 
 ## Bug Fixes
 
-<!-- Here goes notable bug fixes that are worth a special mention or explanation -->
+- The code example extractor relied on a bug in the upstream sybil project that was recently fixed, thus our code also needed to be fixed.
 
 ### Cookiecutter template
 

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -64,7 +64,7 @@ model = []
 extra-lint-examples = [
   "pylint >= 2.17.3, < 4",
   "pytest >= 7.3.0, < 9",
-  "sybil >= 6.1.0, < 7",
+  "sybil >= 6.1.1, < 7",
 ]
 dev-flake8 = [
   "flake8 == 7.0.0",
@@ -107,7 +107,7 @@ dev-pytest = [
   "pylint == 3.1.0",       # We need this to check for the examples
   "cookiecutter == 2.6.0", # For checking the cookiecutter scripts
   "jinja2 == 3.1.3",       # For checking the cookiecutter scripts
-  "sybil == 6.1.0",        # Should be consistent with the extra-lint-examples dependency
+  "sybil >= 6.1.1, < 7",   # Should be consistent with the extra-lint-examples dependency
 ]
 dev = [
   "frequenz-repo-config[dev-mkdocs,dev-flake8,dev-formatting,dev-mkdocs,dev-mypy,dev-noxfile,dev-pylint,dev-pytest]",

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -64,7 +64,7 @@ model = []
 extra-lint-examples = [
   "pylint >= 2.17.3, < 4",
   "pytest >= 7.3.0, < 9",
-  "sybil >= 5.0.3, < 6.1",
+  "sybil >= 6.1.0, < 7",
 ]
 dev-flake8 = [
   "flake8 == 7.0.0",
@@ -107,7 +107,7 @@ dev-pytest = [
   "pylint == 3.1.0",       # We need this to check for the examples
   "cookiecutter == 2.6.0", # For checking the cookiecutter scripts
   "jinja2 == 3.1.3",       # For checking the cookiecutter scripts
-  "sybil == 6.0.3",        # Should be consistent with the extra-lint-examples dependency
+  "sybil == 6.1.0",        # Should be consistent with the extra-lint-examples dependency
 ]
 dev = [
   "frequenz-repo-config[dev-mkdocs,dev-flake8,dev-formatting,dev-mkdocs,dev-mypy,dev-noxfile,dev-pylint,dev-pytest]",

--- a/src/frequenz/repo/config/pytest/examples.py
+++ b/src/frequenz/repo/config/pytest/examples.py
@@ -25,7 +25,6 @@ pytest_collect_file = Sybil(**examples.get_sybil_arguments()).pytest()
 import ast
 import os
 import subprocess
-import textwrap
 from pathlib import Path
 from typing import Any
 
@@ -140,22 +139,11 @@ class _CustomPythonCodeBlockParser(CodeBlockParser):
         )
         imports_code = "\n".join(import_header)
 
-        # Dedent the code example
-        # There is also example.parsed that is already prepared, but it has
-        # empty lines stripped and thus fucks up the line numbers.
-        example_code = textwrap.dedent(
-            example.document.text[example.start : example.end]
-        )
-        # Remove code fences (first/last line)
-        example_code = example_code[
-            example_code.find("\n") + 1 : example_code.rfind("\n")
-        ]
-
         example_with_imports = _FORMAT_STRING.format(
             disable_pylint=_PYLINT_DISABLE_COMMENT.format("disable"),
             imports=imports_code,
             enable_pylint=_PYLINT_DISABLE_COMMENT.format("enable"),
-            code=example_code,
+            code=example.parsed,
         )
 
         # Make sure the line numbers are correct

--- a/src/frequenz/repo/config/pytest/examples.py
+++ b/src/frequenz/repo/config/pytest/examples.py
@@ -31,7 +31,8 @@ from typing import Any
 
 from sybil import Example
 from sybil.evaluators.python import pad
-from sybil.parsers.myst import CodeBlockParser
+from sybil.parsers.myst import CodeBlockParser, PythonCodeBlockParser
+from sybil.typing import Evaluator
 
 _PYLINT_DISABLE_COMMENT = (
     "# pylint: {}=unused-import,wildcard-import,unused-wildcard-import"
@@ -44,22 +45,6 @@ _FORMAT_STRING = """
 {enable_pylint}
 
 {code}"""
-
-
-def get_sybil_arguments() -> dict[str, Any]:
-    """Get the arguments to pass when instantiating the Sybil object to lint docs examples.
-
-    Returns:
-        The arguments to pass when instantiating the Sybil object.
-    """
-    return {
-        "parsers": [_CustomPythonCodeBlockParser()],
-        "patterns": ["*.py"],
-        # This is a hack because Sybil seems to have issues with `__init__.py` files.
-        # See https://github.com/frequenz-floss/frequenz-repo-config-python/issues/113
-        # for details
-        "excludes": ["__init__.py"],
-    }
 
 
 def _get_import_statements(code: str) -> list[str]:
@@ -134,9 +119,9 @@ class _CustomPythonCodeBlockParser(CodeBlockParser):
     Pylint warnings which are unimportant for code examples are disabled.
     """
 
-    def __init__(self) -> None:
+    def __init__(self, language: str | None = None, _: Evaluator | None = None) -> None:
         """Initialize the parser."""
-        super().__init__("python")
+        super().__init__(language, self.evaluate)
 
     def evaluate(self, example: Example) -> None | str:
         """Validate the extracted code example using pylint.
@@ -234,3 +219,27 @@ def _validate_with_pylint(
         return output.splitlines()
 
     return []
+
+
+class MyPythonCodeBlockParser(PythonCodeBlockParser):
+    """Custom Python code block parser that uses the custom code block parser."""
+
+    codeblock_parser_class: type[_CustomPythonCodeBlockParser] = (
+        _CustomPythonCodeBlockParser
+    )
+
+
+def get_sybil_arguments() -> dict[str, Any]:
+    """Get the arguments to pass when instantiating the Sybil object to lint docs examples.
+
+    Returns:
+        The arguments to pass when instantiating the Sybil object.
+    """
+    return {
+        "parsers": [MyPythonCodeBlockParser()],
+        "patterns": ["*.py"],
+        # This is a hack because Sybil seems to have issues with `__init__.py` files.
+        # See https://github.com/frequenz-floss/frequenz-repo-config-python/issues/113
+        # for details
+        "excludes": ["__init__.py"],
+    }

--- a/src/frequenz/repo/config/pytest/examples.py
+++ b/src/frequenz/repo/config/pytest/examples.py
@@ -31,7 +31,7 @@ from typing import Any
 
 from sybil import Example
 from sybil.evaluators.python import pad
-from sybil.parsers.myst import CodeBlockParser, PythonCodeBlockParser
+from sybil.parsers.markdown import CodeBlockParser, PythonCodeBlockParser
 from sybil.typing import Evaluator
 
 _PYLINT_DISABLE_COMMENT = (

--- a/src/frequenz/repo/config/pytest/examples.py
+++ b/src/frequenz/repo/config/pytest/examples.py
@@ -146,8 +146,10 @@ class _CustomPythonCodeBlockParser(CodeBlockParser):
         example_code = textwrap.dedent(
             example.document.text[example.start : example.end]
         )
-        # Remove first line (the line with the triple backticks)
-        example_code = example_code[example_code.find("\n") + 1 :]
+        # Remove code fences (first/last line)
+        example_code = example_code[
+            example_code.find("\n") + 1 : example_code.rfind("\n")
+        ]
 
         example_with_imports = _FORMAT_STRING.format(
             disable_pylint=_PYLINT_DISABLE_COMMENT.format("disable"),

